### ISSUE TITLE
Softmax error

### ISSUE THREAD

**Issue by kobefeng123** [vlfeat/matconvnet#229]
Hi,

I am trying to fine tune the vgg-16 network for VOC 2012 dataset (which has 20 categories), but I am encountering this error:

`Error using  .* `
`Matrix dimensions must agree.`

`Error in vl_nnsoftmax (line 30)`
`Y = Y .* bsxfun(@minus, dzdY, sum(dzdY .* Y, 3)) ;`

`Error in vl_simplenn (line 359)`
       `res(i).dzdx = vl_nnsoftmax(res(i).x, res(i+1).dzdx) ;`

`Error in demo (line 44)`
`res = vl_simplenn(net, im, dzdy, res);`

Actually my test code is quite simple:

`batchsize = 10;`
`net = load('./data/imagenet-vgg-f.mat');`
`im = single(rand(224,224,3,batchsize));`
`labels = ceil(rand(1,batchsize)*2);`
`im = gpuArray(im) ;`
`net = vl_simplenn_move(net, 'gpu') ;`
`one = gpuArray(single(1)) ;`
`net.layers{end}.class = labels ;`
`dzdy = one;`
`res = [] ;`
`res = vl_simplenn(net, im, dzdy, res);`

get_testim() gets a batch of test images and labels, and then feeding to vl_simplenn. I have already changed the last fc layers to output 20 feature maps. The sizes of im and labels are:

`size(im)`

`ans =`

  ` 224   224     3    10`

`size(labels)`

`ans =`

     1     10

I notice that in this post https://github.com/vlfeat/matconvnet/issues/166 has a similar problem, but when I tried to reshape the label to 1 * 1 * 1 * N, the error is the same.

When batch size is equal to 1, or change the softmax function to softmaxloss, my program runs fine. 

Any ideas on this problem?

Thanks!

**Comment by noorjugni** [comment]
@vedaldi
Is there any update on this issue? I get the same error even when batch size is 1.
I need to get probabilities at the classification layer in order to plot roc curve. Can we do this with softmaxloss? If yes, how?

**Comment by omriFdna** [comment]
@vedaldi 

Hi,

Did you manage to solve this issue, I am also getting this error when trying to train a net:

Error using  .* 
Matrix dimensions must agree.

Error in vl_nnsoftmax (line 29)
Y = Y .* bsxfun(@minus, dzdY, sum(dzdY .* Y, 3)) ;

It only works when the batch size is one image.

Thanks

**Comment by lenck** [comment]
Hi, you cannot run training through `vl_nnsoftmax`. In that case you would have to generate the exact distribution after softmax and take a negative of it, which would be numerically unstable, so you would use a logarithm which would make it a `vl_nnsoftmaxlogloss` which solves all that for you ;)
